types/time.go

```diff
@@ -1338,6 +1348,41 @@ func checkDateType(t TimeInternal, allowZeroInDate bool) error {
 	return nil
 }
 
+func checkDateTypeInModify(t TimeInternal, sc *variable.StatementContext) error {
+	year, month, day := t.Year(), t.Month(), t.Day()
+	strict, noZeroDate, noZeroInDate, ignoreErr := true, true, true, false
+	if sc != nil && (sc.InInsertStmt || sc.InUpdateOrDeleteStmt) {
+		m := sc.SQLMode
+		strict, noZeroDate, noZeroInDate, ignoreErr = m.HasStrictMode(), m.HasNoZeroDateMode(), m.HasNoZeroInDateMode(), sc.IgnoreErr
+	}
+	if year == 0 && month == 0 && day == 0 {
+		if !noZeroDate {
+			return nil
+		} else if !strict || ignoreErr {
+			sc.AppendWarning(ErrWarnDataOutOfRange)
+		}
+		return nil
```

Comment:
If `noZeroDate` is true, and in strict mode, insert 0000-00-00 returns no error here?

Reply:
no, there should be an error. There seems still something wrong. I'll take a look